### PR TITLE
Add OIDC end session endpoint and custom query params

### PR DESCRIPTION
**Pull request by shawnhankim**
**Issue:**
- https://github.com/nginxinc/nginx-openid-connect/issues/70

**Summary:**
- Added the IdP's end session endpoint to terminate the user session on the IdP's side.
- Added the customizable variable to support different query parameters from each IdP.
- Enhanced RP's callback URI for NGINX to clean cookies and redirect to the OIDC logout landing page.

**Description:**
- Added a map variable of `$oidc_end_session_endpoint` as same as authorization and token endpoints in the `openid_connect_configuration.conf`.
- Added a map variable of `$oidc_logout_landing_page` to determine where to redirect browser after successful logout from the IdP.
- Added a map variable of `$oidc_end_session_query_params` to support different query parameters per each IdP.
- Enhanced `/logout` location:
  - Add query parameters using `$oidc_end_session_query_params` for the `$oidc_end_session_endpoint`.
  - NGINX Plus: cleared tokens.
  - Redirected to the `$oidc_end_session_endpoint` to start ending session in the IdP.
  
- Enhanced `/_logout` location:
  - Redirected by IdP when IdP successfully finished the session.
  - Clean cookies
  - NGINX Plus: Redirect to the `$oidc_logout_landing_page`.

### CODE REVIEW

**Comment by shawnhankim** [comment]
@route443 :
- Thanks for your review in detail for the [PR](https://github.com/nginxinc/nginx-openid-connect/pull/62).
- This PR is to simplify from the previous [PR](https://github.com/nginxinc/nginx-openid-connect/pull/68).
- For you to easily manage this repo to reduce any concerns from the enhancements based on the reviews on the [PR]((https://github.com/nginxinc/nginx-openid-connect/pull/62)), I have divided a big PR into small PRs, and this is one of PRs.
- Let me know if you have any questions.

**Review by route443 — CHANGES_REQUESTED** [review]
Hi Shawn. Thank you for your contribution to our nginx oidc project.
As I said in previous PRs, we are not using Conventional Commits for this project. So I wouldn't use this approach unnecessarily. Please change the commit name.

File: openid_connect_configuration.conf [hunk 1]
```diff
@@ -28,6 +28,27 @@ map $host $oidc_jwt_keyfile {
     default "http://127.0.0.1:8080/auth/realms/master/protocol/openid-connect/certs";
 }
 
+map $host $oidc_logout_endpoint {
+    default "http://127.0.0.1:8080/auth/realms/master/protocol/openid-connect/logout";
+}
+
+map $host $oidc_logout_query_params_option {
```
[the remark this answers is not in the document]
Comment (shawnhankim) [reply openid_connect_configuration.conf:35]: - Extra parameters over the default parameters (that the NJS provided) make sense. However, it is not fully customizable for customer need as the following examples, or there would be unnecessary parameters which cause additional network packet although it is not too big.

- example 1. Default RP-Initiated Logout Parameter
  ```json
  {
      "post_logout_redirect_uri": "$redirect_base/_logout",
      "id_token_hint": "$session_jwt"
  }
  ```

- example 2. AWS Cognito Logout Parameter
  - If the following parameters are just added as extra arguments, then the unnecessary default parameters are sent to the IdP, and `post_logout_redirect_uri` & `logout_uri` values are also redundant.
   ```json
  {
        "client_id"  : "$oidc_client",
        "logout_uri" : "$redirect_base/_logout"
  }
  ```

- example 3. AWS Cognito Logout Parameter to prompt a user to sign in as another user.
  - If the following parameters are just added as extra arguments, then the unnecessary default parameters are sent to the IdP, and `post_logout_redirect_uri` & `redirect_uri` values are also redundant.

  ```json 
  {
      "response_type": "code",
      "client_id"    : "$oidc_client",
      "redirect_uri" : "$redirect_base$redir_location",
      "state"        : "STATE",
      "scope"        : "$oidc_scopes"
  }
  ```

- example 4. Auth0 Logout Parameter
  - If the following parameters are just added as extra arguments, then the unnecessary default parameters are sent to the IdP, and `post_logout_redirect_uri` & `returnTo` values are also redundant.

  ```json
  {
      "client_id": "$oidc_client",
      "returnTo" : "$redirect_base/_logout"
  }
  ```

Let me know if you just want extra parameters although unnecessary parameters can be sent to the IdP. I will address this based on the NGINX Plus Team's decision.

File: openid_connect_configuration.conf [hunk 2]
```diff
@@ -28,6 +28,27 @@ map $host $oidc_jwt_keyfile {
...
+    # 0: default query params for the RP-initiated logout
+    # 1: extra query params is added after the default query params
+    # 2: replace default query params with custom query params
+    default 0;
+}
+
+map $host $oidc_logout_query_params {
```
[the remark this answers is not in the document]
Comment (shawnhankim) [reply openid_connect_configuration.conf:42]: - Yes, I am aware of it.
- I have suggested the naming change in the https://github.com/nginxinc/nginx-openid-connect/issues/77 for consistency between NGINX Plus and NGINX Management Suite.
- I wanted to synchronize the variable name as it has been recently merged if possible between NGINX Plus and [NGINX Management Suite](https://docs.nginx.com/nginx-management-suite/admin-guides/access-control/configure-authentication/#enable-oidc).
  - NGINX Plus OIDC: `$oidc_authz_extra_args` is merged (Dec/8/2022)
  - NGINX Management Suite: `$oidc_authz_query_params` is released (Jul/20/2022)

**Review by shawnhankim — COMMENTED** [review]
@route443 
- Thanks for your review in detail.
- I have addressed some of comments and am going to address rest of them. 
- In the meantime, I would appreciate it if you could review this PR(https://github.com/nginxinc/nginx-openid-connect/issues/77) to see if we can change the variable name for consistency between NGINX Plus and NGINX Management Suite as `$oidc_authz_extra_args` has been recently merged.

File: openid_connect_configuration.conf [hunk 3]
```diff
@@ -44,12 +65,38 @@ map $host $oidc_scopes {
...
+}
+
 map $host $oidc_logout_redirect {
-    # Where to send browser after requesting /logout location. This can be
-    # replaced with a custom logout page, or complete URL.
+    # This is a RP's callback URI which is called by OP after successful logout.
     default "/_logout"; # Built-in, simple logout page
 }
 
+map $host $post_logout_return_uri {
```
[the remark this answers is not in the document]
Comment (shawnhankim) [reply openid_connect_configuration.conf]: Addressed it.

File: openid_connect_configuration.conf [hunk 4]
```diff
@@ -28,6 +28,27 @@ map $host $oidc_jwt_keyfile {
...
+    default 0;
+}
+
+map $host $oidc_logout_query_params {
+    # Each IdP may use different query params of the $oidc_logout_endpoint. For
+    # example, The Amazon Cognito requires `client_id` and `logout_uri`. The
+    # Auth0 requires `client_id` and `returnTo`. If the map value is empty, then
+    # `post_logout_redirect_uri` and `id_token_hint` are used as default query
+    # params for most of IdPs like AzureAD/Okta/Keycloak/OneLogin/PingIdentity.
+    default "";
```
[the remark this answers is not in the document]
Comment (shawnhankim) [reply openid_connect_configuration.conf:48]: - Yes, it make sense to be more flexible if we provide default value as: `post_logout_redirect_uri=$redirect_base/_logout&id_token_hint=$session_jwt`
  - Then, we don't need 3 options using `$oidc_logout_query_params_option`, and `query params` would be better than `extra args`.
- In the NGINX Management Suite, the variable name of `$session_jwt` is changed with `$id_token`.
- Do you allow to change the variable name from `$session_jwt` to `$id_token` in this repo?

File: openid_connect_configuration.conf [hunk 5]
```diff
@@ -44,12 +65,38 @@ map $host $oidc_scopes {
...
+map $host $oidc_landing_page {
+    # Where to send browser after successful login. This option is only 
+    # recommended for scenarios where a landing page shows default information
+    # without login, and the RP redirects to the landing page after successful
+    # login from the OP. If this is empty, then the RP redirects to $request_uri.
+    default "";
+    #www.example.com $redirect_base;
+}
+
 map $host $oidc_logout_redirect {
```
[the remark this answers is not in the document]
Comment (shawnhankim) [reply openid_connect_configuration.conf:82]: Addressed!

File: openid_connect_configuration.conf (hunk 1 above)
[the remark this answers is not in the document]
Comment (shawnhankim) [reply openid_connect_configuration.conf:35]: - If we provide `$oidc_logout_query_params` with `default "post_logout_redirect_uri=$redirect_base/_logout&id_token_hint=$session_jwt";`, then customers can flexibly customize the query parameters with either extra arguments or replacing all fields.
- So I am going to provide the default value. Let me know if you change your mind.

File: openid_connect_configuration.conf [hunk 6]
```diff
@@ -44,12 +65,38 @@ map $host $oidc_scopes {
...
-    # Where to send browser after requesting /logout location. This can be
-    # replaced with a custom logout page, or complete URL.
+    # This is a RP's callback URI which is called by OP after successful logout.
     default "/_logout"; # Built-in, simple logout page
 }
 
+map $host $post_logout_return_uri {
+    # Where to send browser after the RP requests /logout to the OP, and after
```
[the remark this answers is not in the document]
Comment (shawnhankim) [reply openid_connect_configuration.conf]: Thanks for your suggestion. Consolidated this into `$oidc_logout_landing_page` based on [the review](https://github.com/nginxinc/nginx-openid-connect/pull/72/files#r1058744406).

File: openid_connect_configuration.conf [hunk 7]
```diff
@@ -44,12 +65,38 @@ map $host $oidc_scopes {
...
+    # the RP (/_logout) is called by the OP and cleans cookies. The following
+    # examples can be replaced with a custom logout page, or a complete URL.
+    # If this is empty, then the RP redirects to $request_uri.
+
+    default "";
+
+    # Enable if you want to redirect to the landing page
```
[the remark this answers is not in the document]
Comment (shawnhankim) [reply openid_connect_configuration.conf]: Thanks! Addressed this into `$oidc_logout_landing_page`.

File: openid_connect_configuration.conf [hunk 8]
```diff
@@ -44,12 +65,38 @@ map $host $oidc_scopes {
     default "openid+profile+email+offline_access";
 }
 
+map $host $oidc_landing_page {
+    # Where to send browser after successful login. This option is only
```
[the remark this answers is not in the document]
Comment (shawnhankim) [reply openid_connect_configuration.conf]: Yeah! I will address this with the other PR.

File: openid_connect.server_conf [hunk 9]
```diff
@@ -67,16 +67,30 @@
...
+
+        # Enable following directives, and disable the oidc.redirectPostLogout()
+        # if you want to test a Built-in simple logout page before production.
         default_type text/plain;
         return 200 "Logged out\n";
+
+        # Enable oidc.redirectPostLogout(), and disable the above built-in logout
+        # page if you want to redirect to either the landing page or custom
+        # logout page using the map of $post_logout_return_uri.
+        #js_content oidc.redirectPostLogout;
```
[the remark this answers is not in the document]
Comment (shawnhankim) [reply openid_connect.server_conf]: Addressed it. But, the cookies are cleared here and redirect browser to the `$oidc_logout_landing_page` as this location is called by IdP after successful logout from the IdP. Let me know if you have other thoughts.

File: openid_connect.js [hunk 10]
```diff
@@ -253,11 +256,39 @@ function validateIdToken(r) {
...
+    r.variables.refresh_token = '-';
+    r.return(302, r.variables.oidc_logout_endpoint + queryParams);
+}
+
+// Redirect URI after logged-out from the OP.
+function redirectPostLogout(r) {
+    if (r.variables.post_logout_return_uri) {
+        r.return(302, r.variables.post_logout_return_uri);
+    } else {
+        r.return(302, r.variables.redirect_base + r.variables.cookie_auth_redir);
```
[the remark this answers is not in the document]
Comment (shawnhankim) [reply openid_connect.js]: This was to follow the existing pattern. But, I have revised this as I addressed the idea of `$oidc_logout_landing_page`. Thanks!

File: openid_connect_configuration.conf (hunk 2 above)
[the remark this answers is not in the document]
Comment (shawnhankim) [reply openid_connect_configuration.conf:35]: Also, I wanted to follow the naming of `query params` like the [OpenAPI spec](https://swagger.io/docs/specification/describing-parameters/) and [IETF HTTP Spec](https://www.ietf.org/archive/id/draft-ietf-httpbis-safe-method-w-body-02.html) instead of `args`.
Please let me know what you want.

File: openid_connect_configuration.conf [hunk 11]
```diff
@@ -28,6 +28,18 @@ map $host $oidc_jwt_keyfile {
...
 
+map $host $oidc_logout_endpoint {
+    default "http://127.0.0.1:8080/auth/realms/master/protocol/openid-connect/logout";
+}
+
+map $host $oidc_logout_query_params {
+    # Each IdP may use different query params of the $oidc_logout_endpoint. For
+    # example, Amazon Cognito requires `client_id` and `logout_uri`, and Auth0
+    # requires `client_id` and `returnTo` instead of the default query params.
+    default "post_logout_redirect_uri=$redirect_base/_logout&id_token_hint=$session_jwt";
```
Comment (route443) [inline openid_connect_configuration.conf:39]: I think this approach can still confuse the inexperienced user:
1. We cannot replace the `post_logout_redirect_uri` parameter with something of our own, because the `_logout` location will perform additional functions.
2. We already have 2 similar parameters: `post_logout_redirect_uri` and `$oidc_logout_landing_page`.

Perhaps it makes sense to hide `post_logout_redirect_uri=$redirect_base/_logout&`.

**Review by route443 — CHANGES_REQUESTED** [review]

File: openid_connect_configuration.conf (hunk 2 above)
[the remark this answers is not in the document]
Comment (shawnhankim) [reply openid_connect_configuration.conf]: - It makes sense to use `extra_query_params` rather than `extra_args` although there is no required query parameters declared for the `$oidc_end_session_endpoint`.
- FYI. The required query parameters declared in `authZArgs` for the authz endpoint are not fully enough to be reconfigurable by customers unless they change codes. Hence, we can think of how the required fields' value can be configurable in the future. (Therefore, I used three ways of`extra`, `replace` and `built-in`. )

To sum up: changed it as `extra_query_params` at this moment.

**Review by shawnhankim — COMMENTED** [review]
Thanks for your review. Revised and added some one cent thoughts.

File: openid_connect_configuration.conf (hunk 2 above)
[the remark this answers is not in the document]
Comment (route443) [reply openid_connect_configuration.conf]: It looks like you've lost the thread of the discussion:

1. In the original iteration, you proposed the `$oidc_logout_query_params` var, which involved modifying only a subset of all query parameters.
2. For consistency with `$oidc_authz_extra_args`, I suggested to use to `$oidc_logout_extra_args`.
3. You referred to the fact that in your product (NGINX Management Suite) you already use the `$oidc_authz_query_params` var name and in general the standard says that these arguments are called query parameters.
4. I pointed out at that moment that we are not proposing to modify all the query parameters, but only part of them, then the variable should be called ...extra_query_params.
5. Since in the current iteration you proposed to be able to change ALL query parameters for logout uri, well, call it `...query_params`.

File: openid_connect.js [hunk 12]
```diff
@@ -253,11 +253,26 @@ function validateIdToken(r) {
...
+// An RP requests that the OP log out the end-user by redirecting the end-user's
+// User Agent to the OP's Logout endpoint.
 function logout(r) {
     r.log("OIDC logout for " + r.variables.cookie_auth_token);
-    r.variables.session_jwt = "-";
-    r.variables.refresh_token = "-";
-    r.return(302, r.variables.oidc_logout_redirect);
+    var queryParams = '';
+    if (r.variables.oidc_logout_extra_query_params) {
+        queryParams = '?' + r.variables.oidc_logout_extra_query_params;
```
Comment (route443) [inline openid_connect.js:265]:
```suggestion
        queryParams = '?' + r.variables.oidc_logout_query_params;
```

File: openid_connect.server_conf [hunk 13]
```diff
@@ -1,6 +1,8 @@
     # Advanced configuration START
     set $internal_error_message "NGINX / OpenID Connect login failure\n";
     set $pkce_id "";
+    set $oidc_logout_callback_uri   "$redirect_base/_logout";
```
Comment (route443) [inline openid_connect.server_conf:4]: Introducing a new variable does not make the process any easier.

File: openid_connect_configuration.conf [hunk 14]
```diff
@@ -28,6 +28,18 @@ map $host $oidc_jwt_keyfile {
...
 
+map $host $oidc_end_session_endpoint {
+    default "http://127.0.0.1:8080/auth/realms/master/protocol/openid-connect/logout";
+}
+
+map $host $oidc_logout_extra_query_params {
+    # Each IdP may use different query params of the $oidc_end_session_endpoint.
+    # For example, Amazon Cognito requires `logout_uri=xxx`, and Auth0 requires
+    # `returnTo=xxx` instead of $oidc_logout_callback_param.
+    default "$oidc_logout_callback_param&id_token_hint=$session_jwt";;
```
Comment (route443) [inline openid_connect_configuration.conf:39]: The proposed version did not add simplicity and clarity, but rather the opposite. It's better to return as it was.

**Review by route443 — CHANGES_REQUESTED** [review]

File: openid_connect_configuration.conf [hunk 15]
```diff
@@ -28,6 +28,18 @@ map $host $oidc_jwt_keyfile {
     default "http://127.0.0.1:8080/auth/realms/master/protocol/openid-connect/certs";
 }
 
+map $host $oidc_end_session_endpoint {
+    default "http://127.0.0.1:8080/auth/realms/master/protocol/openid-connect/logout";
+}
+
+map $host $oidc_logout_extra_query_params {
```
[the remark this answers is not in the document]
Comment (shawnhankim) [reply openid_connect_configuration.conf]: Revised it!

File: openid_connect_configuration.conf (hunk 14 above)
Comment (shawnhankim) [reply openid_connect_configuration.conf]: Yeah. Returned as it was!

**Review by shawnhankim — COMMENTED** [review]
Addressed the comments. Let me know if I miss anything. Thanks!

File: openid_connect_configuration.conf (hunk 2 above)
[the remark this answers is not in the document]
Comment (shawnhankim) [reply openid_connect_configuration.conf:35]: Thanks for the notes. I have addressed the comments. Let me know if I miss anything.

**Comment by route443** [comment]
Closed. No response for a long time.
